src/tests/ftest/cart/dual_iface_server.c

```diff
@@ -0,0 +1,434 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * Server utilizing dual interfaces to check connectivity
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <assert.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <semaphore.h>
+#include <cart/api.h>
+#include <cart/types.h>
+
+#include "tests_common.h"
+
+#define MY_BASE 0x010000000
+#define MY_VER  0
+
+#define NUM_SERVER_CTX 8
+
+#define RPC_DECLARE(name)						\
+	CRT_RPC_DECLARE(name, CRT_ISEQ_##name, CRT_OSEQ_##name)		\
+	CRT_RPC_DEFINE(name, CRT_ISEQ_##name, CRT_OSEQ_##name)
+
+enum {
+	RPC_PING = CRT_PROTO_OPC(MY_BASE, MY_VER, 0),
+	RPC_SHUTDOWN
+} rpc_id_t;
+
+#define CRT_ISEQ_RPC_PING	/* input fields */		 \
+	((uint64_t)		(src_rank)		CRT_VAR) \
+	((uint64_t)		(dst_tag)		CRT_VAR)
+
+#define CRT_OSEQ_RPC_PING	/* output fields */		 \
+	((uint64_t)		(rc)			CRT_VAR)
+
+#define CRT_ISEQ_RPC_SHUTDOWN	/* input fields */		 \
+	((uint64_t)		(field)			CRT_VAR)
+
+#define CRT_OSEQ_RPC_SHUTDOWN	/* output fields */		 \
+	((uint64_t)		(field)			CRT_VAR)
+
+static int handler_ping(crt_rpc_t *rpc);
+static int handler_shutdown(crt_rpc_t *rpc);
+
+RPC_DECLARE(RPC_PING);
+RPC_DECLARE(RPC_SHUTDOWN);
+
+struct crt_proto_rpc_format my_proto_rpc_fmt[] = {
+	{
+		.prf_flags	= 0,
+		.prf_req_fmt	= &CQF_RPC_PING,
+		.prf_hdlr	= (void *)handler_ping,
+		.prf_co_ops	= NULL,
+	}, {
+		.prf_flags	= 0,
+		.prf_req_fmt	= &CQF_RPC_SHUTDOWN,
+		.prf_hdlr	= (void *)handler_shutdown,
+		.prf_co_ops	= NULL,
+	}
+};
+
+struct crt_proto_format my_proto_fmt = {
+	.cpf_name = "my-proto",
+	.cpf_ver = MY_VER,
+	.cpf_count = ARRAY_SIZE(my_proto_rpc_fmt),
+	.cpf_prf = &my_proto_rpc_fmt[0],
+	.cpf_base = MY_BASE,
+};
+
+static int
+handler_ping(crt_rpc_t *rpc)
+{
+	struct RPC_PING_in	*input;
+	struct RPC_PING_out	*output;
+	int			my_tag;
+	d_rank_t		hdr_dst_rank;
+	uint32_t		hdr_dst_tag;
+	d_rank_t		hdr_src_rank;
+	int			rc = 0;
+
+	input = crt_req_get(rpc);
+	output = crt_reply_get(rpc);
+
+	rc = crt_req_src_rank_get(rpc, &hdr_src_rank);
+	D_ASSERTF(rc == 0, "crt_req_src_rank_get() failed; rc=%d\n", rc);
+
+	rc = crt_req_dst_rank_get(rpc, &hdr_dst_rank);
+	D_ASSERTF(rc == 0, "crt_req_dst_rank_get() failed; rc=%d\n", rc);
+
+	rc = crt_req_dst_tag_get(rpc, &hdr_dst_tag);
+	D_ASSERTF(rc == 0, "crt_req_dst_tag_get() failed; rc=%d\n", rc);
+
+	crt_context_idx(rpc->cr_ctx, &my_tag);
+
+	if (my_tag != input->dst_tag || my_tag != hdr_dst_tag) {
+		D_ERROR("Incorrect tag Expected %lu got %d (hdr=%d)\n",
+			input->dst_tag, my_tag, hdr_dst_tag);
+		assert(0);
+	}
+	if (hdr_src_rank != input->src_rank) {
+		D_ERROR("Expected %lu got %d\n", input->src_rank, hdr_src_rank);
+		rc = -DER_INVAL;
+	}
+
+	output->rc = rc;
+	crt_reply_send(rpc);
+
+	return 0;
+}
+
+static void
+rpc_handle_reply(const struct crt_cb_info *info)
+{
+	sem_t	*sem;
+
+	D_ASSERTF(info->cci_rc == 0, "rpc response failed. rc: %d\n",
+		  info->cci_rc);
+
+	sem = (sem_t *)info->cci_arg;
+	sem_post(sem);
```

Comment:
What happens if the first cb comes after the it has timed out in the main function so there isn't an error.  Then it will post the semaphore, but that semaphore is also used for the second cb, thus creating a race condition.  Need to abort the first rpc before moving to second rpc.

Reply:
not sure i follow what is first vs second rpc here. each server will only send 1 rpc to another server

Reply:
I see 2 calls to crt_rpc_send specifying this function as the call back.  Each function uses the same local semaphore for synchronization.

Reply:
ok, yes there are 2 rpcs  - one is ping, one is shutdown.
both rpcs are seralized and we assert on any error of the rpc, so we will not hit the race condition with first rpc timing out (as second in such case wont even get invoked)